utils/utf8.cc

```diff
@@ -56,22 +56,30 @@ namespace utils {
 
 namespace utf8 {
 
-// 3x faster than boost utf_to_utf
-static inline std::optional<size_t> validate_naive(const uint8_t *data, size_t len) {
-    size_t pos = 0;
-
-    while (len) {
-        size_t bytes;
-        const uint8_t byte1 = data[0];
+struct codepoint_status {
+    size_t bytes_validated;
+    bool error;
+    uint8_t more_bytes_needed;
+};
 
+static
+codepoint_status
+inline
+evaluate_codepoint(const uint8_t* data, size_t len) {
+    auto byte1 = data[0];
+    static uint8_t len_from_first_nibble[16] = { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 3, 4 };
+    auto codepoint_len = len_from_first_nibble[byte1 >> 4];
+    if (codepoint_len > len) {
+	return codepoint_status{.more_bytes_needed = uint8_t(codepoint_len - len)};
+    } else {
         if (byte1 <= 0x7F) {
             // 00..7F
-            bytes = 1;
-        } else if (len >= 2 && byte1 >= 0xC2 && byte1 <= 0xDF &&
+            return codepoint_status{.bytes_validated = codepoint_len};
+        } else if (byte1 >= 0xC2 && byte1 <= 0xDF &&
                 (int8_t)data[1] <= (int8_t)0xBF) {
```

Comment:
Shouldn't we also check (int8_t)data[1] >= 0x80?

Reply:
There is no (int8_t)data[1] that can be < (int8_t)0x80 - that's the minimum int8_t value. A nice trick by the original author.

In any case it is not related to this refactoting.

Reply:
Nice trick indeed.